uniffi_bindgen/src/interface/types/dispatch.rs

```diff
@@ -0,0 +1,196 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Implement type behavior using the dispatch pattern
+//!
+//! There are a couple issues with implementing type behavior:
+//!
+//!    - `Type` has a lot of variants.  It can be hard to implement traits on `Type` when each
+//!    function needs to match against them all.
+//!    - A secondary issue is that there are types that correspond to `Type` variants (Record,
+//!      Object, CallbackInterface, etc.)  It would be nice for these types to share behavior with
+//!      their `Type` variant.
+//!
+//! This module helps solve both issues using the dispatch pattern and macros:
+//!
+//!  - Define a set of handler structs.  Each one will handle some subset of `Type` variants
+//!    (usually 1).
+//!  - Define a macro that can dispatch function calls for `Type` to one of the handler structs.
+//!  - Wrap traits with the `type_dispatch!` macro, defined here.  `type_dispatch!` derives a trait
+//!    impl for `Type`, `Record`, `Object`, `CallbackInterface`, etc. by dispatching the calls to
+//!    the appropriate handler type.  See `CodeType` and `KotlinCodeType` for an example.
+
+use super::Type;
+
+// Dispatch handler for simple types that correspond to standard types on the target language.
+// It's easier to implement traits for all of these together than with separate structs.
+pub enum SimpleTypeHandler {
+    UInt8,
+    Int8,
+    UInt16,
+    Int16,
+    UInt32,
+    Int32,
+    UInt64,
+    Int64,
+    Float32,
+    Float64,
+    Boolean,
+    String,
+}
+
+// Other variants are mapped 1-1 to structs.
+//
+// When handlers need to reference names and other data from the `Type` instance, we use the `'d`
+// lifetime.  This means that these references are only valid for the dispatched call.
+pub struct RecordTypeHandler<'d> {
```

Comment:
FWIW, I didn't find the word "handler" very meaningful on my first read-through. I don't object to it, it just didn't really do anything to help me understand what was happening. I wonder if something like `RecordTypeDispatcher` work here, leaning in a little bit more to the explicit use of dispatch pattern and the name of this module.

(I'm not willing to put up even a token resistance to the name "handler" if you're happy with it, just sharing my initial impressions)

Reply:
I'm happy to change the name, but I'd prefer something other than dispatcher.  For me "dispatcher" is the thing that routes the method call , the macro in this case, not the thing that the call is routed to.